ipq40xx: Add support for 8devices Habanero DVK

robimarko opened the pull request:
This patch adds support for the 8devices Habanero development board.

Specs are:
CPU: QCA IPQ4019
RAM: DDR3L 512MB
Storage: 32MB SPI-NOR and optional Parallel SLC NAND(Some boards ship with it and some without)
WLAN1: 2.4 GHz built into IPQ4019 (802.11n) 2x2
WLAN2: 5 GHz built into IPO4019 (802.11ac Wawe-2) 2x2
Ethernet: 5x Gbit LAN (QCA 8075)
USB: 1x USB 2.0 and 1x USB 3.0 (Both built into IPQ4019)
MicroSD slot (Uses SD controller built into IPQ4019)
SDIO3.0/EMMC slot (Uses the same SD controller)
Mini PCI-E Gen 2.0 slot (Built into IPQ4019)
5x LEDs (4 GPIO controllable)
2x Pushbutton (1 is connected to GPIO, other to SoC reset)
LCD ZIF socket (Uses the LCD controller built into IPQ4019 which has no driver support)
1x UART 115200 rate on J18

2x breakout development headers
12V DC Jack for power
DIP switch for bootstrap configuration

Installation instructions:
Since boards ship with vendors fork of OpenWrt sysupgrade can be used.

Signed-off-by: Robert Marko <robimarko@gmail.com>

adrianschmutzler commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -24,6 +24,7 @@ endef
 #
 
 ALLWIFIBOARDS:= \
+  8dev_habanero-dvk \
```
Fix indent.

robimarko replied:
Will do

robimarko commented:
@adrianschmutzler Your remarks have been fixed

mantas-p commented:
There has been one more board-data update [1], it would be great if you'd pull that in.


[1] https://github.com/8devices/openwrt-8devices/commit/5651870da82a30ab9d8c7f42c258278530f88428

robimarko commented:
Sure, will pull that

adrianschmutzler commented:
Does this device have a MAC address on label/sticker/case/box?

robimarko commented:
Yes, MAC addresses are on the label for 2.4 and 5 GHz radios.
Why?

adrianschmutzler commented:
> Yes, MAC addresses are on the label for 2.4 and 5 GHz radios.
> Why?

I've recently added some patches that enable to store the label MAC address for devices in OpenWrt. Many devices bear a label with a single MAC address on them, which is used by several communities to identify devices in a network.

However, if your label shows two addresses, one would have to choose one of them arbitrarily.

robimarko commented:
@adrianschmutzler Added label-mac-device.
Anything more you spotted?

adrianschmutzler commented:
@robimarko 

I'm not sure label-mac-device will work here. This would require a mac-address or local-mac-address subnode of wlan0 in DTS.

In this case, the following should return a MAC address:
```
. /lib/functions/system.sh
get_mac_label
```

If label-mac-device is not working, you can add the label MAC address in 02_network, as I did with FritzBox 4040 in this patch:
https://patchwork.ozlabs.org/patch/1166089/

dengqf6 commented on target/linux/ipq40xx/config-4.19:
```diff
@@ -283,6 +284,14 @@ CONFIG_MFD_SYSCON=y
+CONFIG_MMC_SDHCI_MSM=y
+# CONFIG_MMC_SDHCI_PCI is not set
+CONFIG_MMC_SDHCI_PLTFM=y
+# CONFIG_MMC_TIFM_SD is not set
```
Is it possible to build the SDHCI as a module instead?

dengqf6 replied:
Making this built-in will cause unnecessary kernel bloat.
You can make a separate package just like `kmod-sdhci-mt7620`

robimarko replied:
Do we really need another kmod?
There is no practical benefit to it, ipq40xx is not resource-constrained and few kilobytes are not an issue here.
MT7620 is much more space-constrained usually.

robimarko commented:
@adrianschmutzler You were right, label will not work in DTS.
But there is also and issue with using 02_network script.
Its executed too early and since radios are probed and registered only at the end of boot there is no sysfs entry for them.

> cat: can't open '/sys/class/net/wlan0/address': No such file or director

I can use ethernet address but there are not listed on the sticker at all

adrianschmutzler commented:
You need a different path for WiFi:

label_mac=$(cat /sys/class/ieee80211/phy0/macaddress)
(or phy1)

e.g. here
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/base-files/etc/board.d/02_network#L619

I tested this once on an ath79 device (I do not remember which any more), and then it worked.

robimarko commented:
Well, it won't work either, its the same deal.
As radios are probed and registered last in the boot process there is no wiphy registered.

> cat: can't open '/sys/class/ieee80211/phy0/macaddress': No such file or director

adrianschmutzler commented:
Interesting. Just checked again on my Archer C7 v5 (ath79).
There, it works.

I just put the following in 02_network:
+               cat /sys/class/ieee80211/phy0/macaddress > /etc/testphy0.txt
+               cat /sys/class/ieee80211/phy1/macaddress > /etc/testphy1.txt
+               cat /sys/class/net/eth0/address > /etc/testeth0.txt
+               cat /sys/class/net/eth1/address > /etc/testeth1.txt

and then flashed the new image with sysupgrade -n.

I found a MAC address in every file except for eth1 (which is not present and was meant as blind test).

So, it looks like phy probing in ipq40xx takes place later than in ath79?

robimarko commented:
Hm, it looks like it.
In ipq40xx it's pretty much the last thing to happen because ath10k kmod is one of the last ones to modprobe

adrianschmutzler commented:
Note that the C7v5 also has an ath10k radio.

The only ipq40xx device I have is the FritzBox 4040. I will have a look whether I can reproduce the situation there.
A workaround could be to just put `uci set system.@system[0].label_macaddr=$(cat ...)` somewhere late enough in the boot process, as that's what setting label_mac is essentially doing at the end. But I do not really like that ...

robimarko commented:
Yeah, I checked and it has ath10k radio but for whatever reason, probing happens really late on ipq40xx.
It would be great if you can test it on the Fritzbox

adrianschmutzler commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -130,6 +130,7 @@ case "$FIRMWARE" in
 	;;
 "ath10k/pre-cal-ahb-a000000.wifi.bin")
 	case "$board" in
+	8dev,habanero-dvk |\
```
Hi, since you are using 11-ath10k-caldata  here, I wanted to kindly ask you whether you could test my patchset for caldata extraction unification in PR #2464 on this device to cover ipq40xx.

You would just need the first four patches from my PR.

I'm just asking for a favor, so feel free to decline this request and just mark it as resolved then.

robimarko replied:
Sorry, I would gladly test but I am spread so thin with time currently.

adrianschmutzler commented:
> Yeah, I checked and it has ath10k radio but for whatever reason, probing happens really late on ipq40xx.
> It would be great if you can test it on the Fritzbox

On the FritzBox, I can access the valid MAC address from phy0/phy1 during 02_network execution. Don't really know what's different with your device in this case...

robimarko commented:
I don't either, I tried again and during 02_network I can't access the MAC but after probing I can.

adrianschmutzler commented:
Well, you use the ipq-wifi firmware, and the FritzBox doesn't. Maybe that leads to later probing?

robimarko commented:
It shouldn't as all it does is replace the board-2.bin with the custom one during build time.
So it should not affect runtime at all

adrianschmutzler commented:
Okay, then I'm out of ideas, especially since I'm not really used to ipq40xx target anyway.
So, this device might have to stay without label MAC address ...

If you find some spare time, maybe check for the other devices you want to add (unless you can use DTS-based approach there ...).

robimarko commented:
Yeah, will try this on couple other ipq40xx devices I have.

robimarko commented:
@chunkeey Any chance of merging this?

robimarko commented:
SDHCI node was finally accepted upstream and will be in kernel 5.5

https://git.kernel.org/pub/scm/linux/kernel/git/qcom/linux.git/commit/?id=04b3b72b5b8fdb883bfdc619cb29b03641b1cc6a

adrianschmutzler commented:
Just FYI: phy0/phy1 MAC address is also accessible during 02_network on Netgear EX6150v2 (ipq4018 as the FritzBox).

robimarko commented:
Well, then I really don't understand why does it not work on Habanero.

adrianschmutzler commented:
Maybe it's 4018 vs. 4019. Maybe it's something completely different.
Unfortunately, I'm not too familiar with ipq40xx target and also do not have the necessary devices, so I can't really be any help beyond the obvious comment above.
Fortunately, having label MAC address isn't utterly important ;-)

robimarko commented:
Refreshed to apply to current master

adrianschmutzler commented:
@robimarko During other projects I learned that phyX setup is asynchronous. However, we actually do not have to rely on phyX for the label_mac address, but can just use the caldata.
So, depending on which phy is the right one, you can just set
label_mac=$(mtd_get_mac_binary "ART" 0x1006)
or
label_mac=$(mtd_get_mac_binary "ART" 0x5006)

robimarko commented:
Awesome, I will then apply that to this PR.
That should resolve the last remarks.

robimarko commented:
And, I pushed the rebased version with label MAC for 2.4GHz radio included.
Hopefully, it now can be merged.

ynezz commented on target/linux/ipq40xx/config-4.19:
```diff
@@ -66,6 +66,7 @@ CONFIG_BLK_DEV_LOOP=y
 CONFIG_BLK_MQ_PCI=y
 CONFIG_BOUNCE=y
 # CONFIG_CACHE_L2X0 is not set
+CONFIG_CC_HAS_ASM_GOTO=y
```
This is runtime option, so it shouldn't be set.

robimarko replied:
I did not set it manually, it was set by make kernel_menuconfig when I was enabling SDHCI and LDO driver.
I have also seen it on other targets being enabled lately.

ynezz replied:
>  I have also seen it on other targets being enabled lately.

It doesn't mean, that it's correct.

robimarko replied:
I can manually remove it then, it looks like simple kernel config refresh will enable it.

ynezz replied:
Yes, we need to [live with that](https://bugs.openwrt.org/index.php?do=details&task_id=2588) or fix our tools :-)

ynezz commented:
Generally I'm fine with `ipq40xx: Add support for 8devices Habanero DVK`, but I cant say the same about the SD/MMC related patches, because I simply dont like the idea of adding another patches we need to maintain and support.

So I would suggest to make this PR only about adding support for the DVK (I can give ACK to that PR ASAP) and create another PR which would add SD/MMC support even with the v1 of patches sent upstream.

robimarko commented:
Ok, it makes sense.
SDHCI node is already accepted upstream and will be in kernel 5.5.
LDO driver is issue, since I am not the original author and I only partly modified it.
@mantas-p Is the author so I cant submit it upstream.

ynezz commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-habanero-dvk.dts:
```diff
@@ -0,0 +1,360 @@
+	status = "okay";
+	cs-gpios = <&tlmm 12 GPIO_ACTIVE_HIGH>;
+
+	m25p80@0 {
```
flash@0

ynezz commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-habanero-dvk.dts:
```diff
@@ -0,0 +1,360 @@
+				label = "cfg";
+				reg = <0x00180000 0x00040000>;
+			};
+			partition9@1a0000 {
```
@1c0000

adrianschmutzler commented:
If you change anyway, another nitpick: note that the "Add" in commit title after the "ipq40xx:" prefix needs to be lower case ...

ynezz commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-habanero-dvk.dts:
```diff
@@ -0,0 +1,360 @@
+};
+
+/* Some DVK boards ship with NAND soldered
+ *while others dont */
```
I would remove the comment, you're enabling it anyway for all.

robimarko replied:
Yes, but this only enables the controller.
Without adding the nand@0(Or similar) node NAND wont work as CS will be missing.
So I thought that it made sense to note that fact.

ynezz replied:
Ok, then make it shorter and on single line `/* Some DVK boards ships without NAND */` or such.

ynezz commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-habanero-dvk.dts:
```diff
@@ -0,0 +1,360 @@
+&nand {
+	pinctrl-0 = <&nand_pins>;
+	pinctrl-names = "default";
+	status = "okay";
```
Just a nitpick, It's quite common habbit to add `status = "okay";` as a first property in the node, followed by the new line and then rest of the properties.

ynezz commented on target/linux/ipq40xx/patches-4.19/304-spi-nor-enable-4B-opcodes-for-w25q256jv.patch:
```diff
@@ -0,0 +1,12 @@
+ 	{ "w25q80", INFO(0xef5014, 0, 64 * 1024,  16, SECT_4K) },
+ 	{ "w25q80bl", INFO(0xef4014, 0, 64 * 1024,  16, SECT_4K) },
+ 	{ "w25q128", INFO(0xef4018, 0, 64 * 1024, 256, SECT_4K) },
+-	{ "w25q256", INFO(0xef4019, 0, 64 * 1024, 512, SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ) },
```
This needs to go into separate commit with proper description, ideally sent upstream...

robimarko replied:
Well, as far as I understand the reason why this was never upstreamed was that regular w25q256 and w25q256 share the same JEDEC ID so you really cant figure out if its old or JV revision

ynezz replied:
Separate commit please.

chunkeey replied:
5.4 can't come soon enough...

ynezz commented:
> LDO driver is issue, since I am not the original author and I only partly modified it.
> @mantas-p Is the author so I cant submit it upstream.

You can do it and as I know @mantas-p he wont be against, you just need to preserve the authorship, proper `From:` and `Signed-off-by:` etc.

ynezz commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-habanero-dvk.dts:
```diff
@@ -0,0 +1,360 @@
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
```
Replace this boilerplate text with `SPDX-License-Identifier: GPL-2.0-or-later OR
MIT` license, as this seems to [be preferred by the
upstream](https://github.com/openwrt/openwrt/pull/1668#issuecomment-449692907)

ynezz commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-habanero-dvk.dts:
```diff
@@ -0,0 +1,360 @@
+		};
+	};
+
+	gpio-leds {
```
Could you please change this to generic name `leds` as suggested by [DTS Checklist](https://openwrt.org/submitting-patches#dts_checklist)
>  5. The name of a node should reflect the function of the device and not its model. Examples for generic node names can be found in Section 2.2.2 Generic Names Recommendation

ynezz commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-habanero-dvk.dts:
```diff
@@ -0,0 +1,360 @@
+		};
+	};
+
+	gpio-keys {
```
Could you please change this to generic name `keys` as suggested by [DTS Checklist](https://openwrt.org/submitting-patches#dts_checklist)
>  5. The name of a node should reflect the function of the device and not its model. Examples for generic node names can be found in Section 2.2.2 Generic Names Recommendation

ynezz commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-habanero-dvk.dts:
```diff
@@ -0,0 +1,360 @@
+		};
+
+		tcsr@194b000 {
+			/* select hostmode */
```
This comment is needed?

adrianschmutzler commented:
@robimarko You have different mail addresses for "From:" and "Signed-off-by:". That might cause problem when pushing the patch. Despite, please make the "Add" in your commit title lower case.

robimarko commented:
I know, I still need to push the updated branch today.

robimarko commented:
Also, the 4B opcode is waiting for some feedback from upstream.

robimarko commented:
@adrianschmutzler @ynezz I have pushed the updated branch without SD/LDO support and with your remarks addressed except for 4B opcodes.
That is submitted upstream and is waiting for feedback

ynezz commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-habanero-dvk.dts:
```diff
@@ -0,0 +1,332 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/* Copyright (c) 2019, Robert Marko <robimarko@gmail.com> */
```
I don't know the origin of the content, but previously there was LF copyright and it was licensed under MIT...

ynezz commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-habanero-dvk.dts:
```diff
@@ -0,0 +1,332 @@
+	perst-gpio = <&tlmm 38 GPIO_ACTIVE_LOW>;
+	wake-gpio = <&tlmm 50 GPIO_ACTIVE_LOW>;
+
+	/* Free slot for use */
```
Is it really needed?

ynezz commented:
> with your remarks addressed except for 4B opcodes.

This is show stopper. You need to put it into separate commit with a good explanation.

> That is submitted upstream and is waiting for feedback

Nice! Please add a link to the patch into the commit description.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-habanero-dvk.dts:
```diff
@@ -0,0 +1,332 @@
+			panic-indicator;
+		};
+
+		upgrade: upgrade {
```
Didn't see this the first time: I think it's good practice to prefix the node _labels_ for leds with "led_", as those exist in a global context. By this one can immediately understand what he is looking at only from the label.

So, I'd upgrade this to `led_upgrade: upgrade {`, same for status, and update aliases.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-habanero-dvk.dts:
```diff
@@ -0,0 +1,332 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition0@0 {
```
Don't know whether this is different on ipq40xx, but I'd remove the digit before the "@":
partition0@0 => partition@0
partition1@40000 => partition@40000
etc.

robimarko commented:
@ynezz @adrianschmutzler I finally got around to this.
I managed to upstream the NOR 4B, SD LDO driver, SDHCI node, only LDO node is awaiting to be merged upstream.

I also fixed some syntax issues, hopefully its now ready for merging.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-habanero-dvk.dts:
```diff
@@ -0,0 +1,357 @@
+				read-only;
+			};
+			partition@e0000 {
+				label = "APPSBLENV"; /* uboot env*/
```
Nitpick: missing space after "env".

robimarko replied:
Done

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -61,6 +61,18 @@ define Build/SenaoFW
+	$(call Device/FitImageLzma)
+	DEVICE_VENDOR := 8devices
+	DEVICE_MODEL := Habanero DVK
+	DEVICE_DTS := qcom-ipq4019-habanero-dvk
```
Use `SOC := qcom-ipq4019` instead.

robimarko replied:
Done

robimarko commented:
@ynezz I resolved the conflicts for 4.19.
Will add 5.4 support now, so after that can this be finally merged?

robimarko committed:
ipq40xx: add support for 8devices Habanero DVK

This patch adds support for the 8devices Habanero development board.

Specs are:
CPU: QCA IPQ4019
RAM: DDR3L 512MB
Storage: 32MB SPI-NOR and optional Parallel SLC NAND(Some boards ship with it and some without)
WLAN1: 2.4 GHz built into IPQ4019 (802.11n) 2x2
WLAN2: 5 GHz built into IPO4019 (802.11ac Wawe-2) 2x2
Ethernet: 5x Gbit LAN (QCA 8075)
USB: 1x USB 2.0 and 1x USB 3.0 (Both built into IPQ4019)
MicroSD slot (Uses SD controller built into IPQ4019)
SDIO3.0/EMMC slot (Uses the same SD controller)
Mini PCI-E Gen 2.0 slot (Built into IPQ4019)
5x LEDs (4 GPIO controllable)
2x Pushbutton (1 is connected to GPIO, other to SoC reset)
LCD ZIF socket (Uses the LCD controller built into IPQ4019 which has no driver support)
1x UART 115200 rate on J18

2x breakout development headers
12V DC Jack for power
DIP switch for bootstrap configuration

Installation instructions:
Since boards ship with vendors fork of OpenWrt sysupgrade can be used.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
ipq40xx: add IPQ4019 SD/MMC controller support

This commit finally adds support for the built in SD/MMC controller in IPQ4019 SoC.

Controller is supported by the upstream SDHCI-MSM driver with a minor clock setting patch.
Patch is special to the IPQ4019 and cannot be upstreamed.

LDO and SDHCI node are upstreamed, and LDO node is awaiting to be accepted.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
@ynezz 5.4 support added

ynezz commented:
Thanks, pushed.

suyuan168 commented:
usb3: IRQ ss_phy_irq not found
habanero-dvk usb3 speed low
 But there is no matter whether the firmware downloaded from the official or the firmware compiled by yourself. The test USB speed is at most 101 Mbits / sec. This is completely lower than the normal USB3 speed. How is this going? Is there any solution?

robimarko commented:
That warning is expected.
Sorry, but on what are you testing the speed?

suyuan168 commented:
> That warning is expected.
> Sorry, but on what are you testing the speed?

I added a USB network card via USB3. Then physically bridge to lan. AX88179 or any model. Open iperf3 -s in openwrt Open iperf3 -c on the local computer to perform Iperf3 speed measurement.
The speed is very slow. The result does not exceed 100mbit / s,

suyuan168 commented:
[  5]  13.00-14.00  sec  5.41 MBytes  45.4 Mbits/sec                  
[  5]  14.00-15.00  sec  5.76 MBytes  48.2 Mbits/sec                  
[  5]  15.00-16.00  sec  17.6 MBytes   147 Mbits/sec                  
[  5]  16.00-17.00  sec  3.24 MBytes  27.3 Mbits/sec                  
[  5]  17.00-18.00  sec  16.0 MBytes   134 Mbits/sec                  
[  5]  18.00-19.00  sec  8.12 KBytes  66.4 Kbits/sec                  
[  5]  19.00-20.00  sec  1.93 MBytes  16.2 Mbits/sec                  
[  5]  20.00-21.00  sec  11.1 MBytes  93.4 Mbits/sec                  
[  5]  21.00-22.00  sec  23.2 KBytes   190 Kbits/sec                  
[  5]  22.00-23.00  sec  7.14 MBytes  59.8 Mbits/sec                  
[  5]  23.00-24.00  sec  5.07 MBytes  42.6 Mbits/sec                  
[  5]  24.00-25.00  sec  14.8 MBytes   124 Mbits/sec                  
[  5]  25.00-26.00  sec  28.3 KBytes   232 Kbits/sec                  
[  5]  26.00-27.00  sec  10.6 MBytes  88.7 Mbits/sec                  
[  5]  27.00-28.00  sec  4.08 MBytes  34.2 Mbits/sec                  
[  5]  28.00-29.00  sec  5.04 MBytes  42.3 Mbits/sec                  
[  5]  29.00-30.00  sec  20.1 MBytes   169 Mbits/sec                  
[  5]  30.00-31.00  sec  28.3 KBytes   232 Kbits/sec                  
[  5]  31.00-32.00  sec  16.2 MBytes   136 Mbits/sec                  
[  5]  32.00-33.00  sec  28.3 KBytes   232 Kbits/sec                  
[  5]  33.00-34.00  sec  13.8 MBytes   116 Mbits/sec                  
[  5]  34.00-35.00  sec  13.2 MBytes   111 Mbits/sec                  
[  5]  35.00-36.00  sec  15.4 MBytes   129 Mbits/sec                  
[  5]  36.00-37.00  sec  13.5 MBytes   114 Mbits/sec                  
[  5]  37.00-38.00  sec  8.79 MBytes  73.6 Mbits/sec

robimarko commented:
Have you tested 8devices images?

USB networking cards are not really the best way to test USB speeds.
Does the kernel recognize the card as SuperSpeed device?

suyuan168 commented:
Yes. I use the latest image downloaded by 8devices.
And recognized as SuperSpeed USB device.
My AX88179 is recognized as a Gigabit network card.
Anyway, I cannot use USB3 devices normally. Because the speed is always unstable.
Or copy my SSD USB3 copy file.
Still take ipef3 to test the speed.
please help me. anyway. I even feel that the ipq4019's USB3 device itself is very slow?

suyuan168 commented:
> Have you tested 8devices images?
> 
> USB networking cards are not really the best way to test USB speeds.
> Does the kernel recognize the card as SuperSpeed device?

Yes. I use the latest image downloaded by 8devices.
And recognized as SuperSpeed USB device.
My AX88179 is recognized as a Gigabit network card.
Anyway, I cannot use USB3 devices normally. Because the speed is always unstable.
Or copy my SSD USB3 copy file.
Still take ipef3 to test the speed.
please help me. anyway. I even feel that the ipq4019's USB3 device itself is very slow?

suyuan168 commented:
[  518.217903] usb 2-1: new SuperSpeed Gen 1 USB device number 3 using xhci-hcd
[  518.254658] usb 2-1: no of_node; not parsing pinctrl DT
[  518.292144] ax88179_178a 2-1:1.0: no of_node; not parsing pinctrl DT
[  518.692896] ax88179_178a 2-1:1.0 eth2: register 'ax88179_178a' at usb-xhci-hcd.0.auto-1, ASIX AX88179 USB 3.0 Gigabit Ethernet, 00:0e:c6:ab:62:c8

suyuan168 commented:
[    7.638729] random: crng init done
[   10.385667] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[   10.642662] SCSI subsystem initialized
[   10.657486] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[   10.658490] ohci-platform: OHCI generic platform driver
[   10.664016] uhci_hcd: USB Universal Host Controller Interface driver
[   10.668789] ohci-pci: OHCI PCI platform driver
[   10.676746] dwc3-qcom 8af8800.usb3: IRQ hs_phy_irq not found
[   10.678540] dwc3-qcom 8af8800.usb3: IRQ dp_hs_phy_irq not found
[   10.684352] dwc3-qcom 8af8800.usb3: IRQ dm_hs_phy_irq not found
[   10.690066] dwc3-qcom 8af8800.usb3: IRQ ss_phy_irq not found
[   10.701460] dwc3-qcom 60f8800.usb2: IRQ hs_phy_irq not found
[   10.701804] dwc3-qcom 60f8800.usb2: IRQ dp_hs_phy_irq not found
[   10.707470] dwc3-qcom 60f8800.usb2: IRQ dm_hs_phy_irq not found
[   10.713084] dwc3-qcom 60f8800.usb2: IRQ ss_phy_irq not found
[   10.721299] dwc3 8a00000.dwc3: Failed to get clk 'ref': -2
[   10.877810] dwc3 6000000.dwc3: Failed to get clk 'ref': -2
[   11.040932] xhci-hcd xhci-hcd.0.auto: xHCI Host Controller
[   11.040995] xhci-hcd xhci-hcd.0.auto: new USB bus registered, assigned bus number 1
[   11.045549] xhci-hcd xhci-hcd.0.auto: hcc params 0x0228f665 hci version 0x100 quirks 0x0000000002010010
[   11.052998] xhci-hcd xhci-hcd.0.auto: irq 98, io mem 0x08a00000
[   11.063380] hub 1-0:1.0: USB hub found
[   11.068240] hub 1-0:1.0: 1 port detected
[   11.072316] xhci-hcd xhci-hcd.0.auto: xHCI Host Controller
[   11.076059] xhci-hcd xhci-hcd.0.auto: new USB bus registered, assigned bus number 2
[   11.081393] xhci-hcd xhci-hcd.0.auto: Host supports USB 3.0 SuperSpeed
[   11.093744] usb usb2: We don't know the algorithms for LPM for this host, disabling LPM.
[   11.096218] hub 2-0:1.0: USB hub found
[   11.103811] hub 2-0:1.0: 1 port detected

suyuan168 commented:
If the kernel is changed to 5.4.36, I will maintain the speed at 100Mbit / s. But still very slow

suyuan168 commented:
<img width="1042" alt="WX20200520-234749@2x" src="https://user-images.githubusercontent.com/3435645/82467559-5e1edc80-9af4-11ea-8ee4-00983439022a.png">
